[WIP] Add `instance_pools_to_use` parameter to `aws_spot_fleet_request`

zytek opened the pull request:
Fixes #5928

Changes proposed in this pull request:

* add `instance_pools_to_use` parameter to `aws_spot_fleet_request` resource

Output from acceptance testing:

* not run, please assist 

Please comment and review, this is my attempt at working with `terraform` code.

zytek commented:
@bflad I've seen you used different name for this param in ec2_fleet. Should I update this PR with similar change? Any other comments?

bflad commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -623,6 +629,13 @@ func resourceAwsSpotFleetRequestCreate(d *schema.ResourceData, meta interface{})
 		spotFleetConfig.AllocationStrategy = aws.String("lowestPrice")
 	}
 
+	if v, ok := d.GetOk("instance_pools_to_use"); ok {
```
I would recommend checking against the default value and removing our error check here (letting the API return the appropriate error if necessary), e.g.

```go
if v, ok := d.GetOk("instance_pools_to_use_count"); ok && v.(int) != 1 {
  spotFleetConfig.InstancePoolsToUseCount = aws.Int64(int64(v.(int)))
}
```

zytek replied:
@bflad could you elaborate on why you prefer to return errors from API rather than catch them earlier? I am no opposing it, just curious about the rationale. 

Working on updating this PR, I see one acceptance test failing due to spot prices jump, not sure if I should alter test spec or ignore it.

bflad replied:
> could you elaborate on why you prefer to return errors from API rather than catch them earlier? I am not opposing it, just curious about the rationale.

Unless we can generate `plan` time errors (multi-attribute validation in Terraform is a can of worms at the moment), we prefer the API because:
* Upstream API logic may change 
* Upstream API may remove the restriction
* Less code on our end

Which generally boils down to less maintenance burden in this project (someone updating the code then getting a release cut) and for operators (requiring everyone to update their provider). This is the same reasoning why we try to avoid things like EC2 instance type validation, it can be too fast of a moving target or provide little benefit for all the effort.

> Working on updating this PR, I see one acceptance test failing due to spot prices jump, not sure if I should alter test spec or ignore it.

Is it this one? #6031 😄  You can ignore it or potentially look at the forward looking solution mentioned in the comment.

zytek replied:
Thanks for clarification. Yeah, the PR you mentioned is the exact issue I got. ;-)

zytek commented:
Added tests and addressed comments by @bflad 

Two test failures but not related.

Output:
```
➜  terraform-provider-aws git:(spot_fleet_pools) ✗ time make testacc TESTARGS='-run=TestAccAWSSpotFleetRequest_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSSpotFleetRequest_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]


=== RUN   TestAccAWSSpotFleetRequest_associatePublicIpAddress
--- PASS: TestAccAWSSpotFleetRequest_associatePublicIpAddress (261.95s)
=== RUN   TestAccAWSSpotFleetRequest_instanceInterruptionBehavior
--- PASS: TestAccAWSSpotFleetRequest_instanceInterruptionBehavior (262.61s)
=== RUN   TestAccAWSSpotFleetRequest_fleetType
--- PASS: TestAccAWSSpotFleetRequest_fleetType (252.01s)
=== RUN   TestAccAWSSpotFleetRequest_iamInstanceProfileArn
--- PASS: TestAccAWSSpotFleetRequest_iamInstanceProfileArn (274.72s)
=== RUN   TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
--- PASS: TestAccAWSSpotFleetRequest_changePriceForcesNewRequest (532.76s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion (262.78s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList (283.29s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
--- FAIL: TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList (78.48s)
    testing.go:527: Step 0 error: Error applying: 1 error occurred:
        	* aws_spot_fleet_request.foo: 1 error occurred:
        	* aws_spot_fleet_request.foo: Last events: [{
          EventInformation: {
            EventDescription: "m3.large, ami-d0f506b0, Linux/UNIX, us-west-2b, Spot bid price is less than Spot market price $0.0307",
            EventSubType: "launchSpecUnusable"
          },
          EventType: "information",
          Timestamp: 2018-10-02 13:36:49.663 +0000 UTC
        } {
          EventInformation: {
            EventDescription: "m3.large, ami-d0f506b0, Linux/UNIX, us-west-2a, Spot bid price is less than Spot market price $0.0307",
            EventSubType: "launchSpecUnusable"
          },
          EventType: "information",
          Timestamp: 2018-10-02 13:36:49.674 +0000 UTC
        } {
          EventInformation: {
            EventDescription: "All launch specifications are unusable. Please refer to individual information events for more detail.",
            EventSubType: "fleetProgressHalted"
          },
          EventType: "information",
          Timestamp: 2018-10-02 13:36:49.682 +0000 UTC
        }]




=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz (242.33s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet (276.12s)
=== RUN   TestAccAWSSpotFleetRequest_overriddingSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_overriddingSpotPrice (265.47s)
=== RUN   TestAccAWSSpotFleetRequest_withoutSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_withoutSpotPrice (255.23s)
=== RUN   TestAccAWSSpotFleetRequest_diversifiedAllocation
--- PASS: TestAccAWSSpotFleetRequest_diversifiedAllocation (343.92s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstancePools
--- PASS: TestAccAWSSpotFleetRequest_multipleInstancePools (391.19s)
=== RUN   TestAccAWSSpotFleetRequest_withWeightedCapacity
--- PASS: TestAccAWSSpotFleetRequest_withWeightedCapacity (272.87s)
=== RUN   TestAccAWSSpotFleetRequest_withEBSDisk
--- PASS: TestAccAWSSpotFleetRequest_withEBSDisk (243.26s)
=== RUN   TestAccAWSSpotFleetRequest_withTags
--- FAIL: TestAccAWSSpotFleetRequest_withTags (412.63s)
    testing.go:588: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.

        Error: Error applying: 1 error occurred:
        	* aws_spot_fleet_request.foo (destroy): 1 error occurred:
        	* aws_spot_fleet_request.foo: error deleting spot request (sfr-b897cfaa-5808-42af-856c-cfcb3c46fe24): fleet still has (1) running instances
[...]
=== RUN   TestAccAWSSpotFleetRequest_placementTenancy
--- PASS: TestAccAWSSpotFleetRequest_placementTenancy (67.97s)
=== RUN   TestAccAWSSpotFleetRequest_WithELBs
--- PASS: TestAccAWSSpotFleetRequest_WithELBs (318.78s)
=== RUN   TestAccAWSSpotFleetRequest_WithTargetGroups
--- PASS: TestAccAWSSpotFleetRequest_WithTargetGroups (418.43s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	5716.830s
make: *** [testacc] Error 1
make testacc TESTARGS='-run=TestAccAWSSpotFleetRequest_'  30.54s user 6.88s system 0% cpu 1:35:26.04 total
```
